### PR TITLE
Multilanguage

### PR DESCRIPTION
**Pull request by DanBendavid**
modified:   hdate/date.py
	modified:   hdate/htables.py
	modified:   hdate/zmanim.py
	modified:   tests/test_hdate.py
	modified:   tests/test_hdate_api.py
	modified:   tests/test_hdate_string.py

# Description
Adding a parameter to handle multiple language : language="hebrew" instead of  hebrew=True or hebrew=False for english
language available : french, english, hebrew
 
# Closes issue(s)

  Fixes: #<issue number>

# Changes included (select only one)

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [X] Breaking change (change that is not backwards-compatible)

# Checklist

- [x] Code passes tox
- Refactoring of Holiday list with the addition of holiday_description covering french, english and hebrew

### CODE REVIEW

**Comment by tsvi** [comment]
You'll need to fix the lint errors. Also please use language instead of lang.

File: hdate/zmanim.py [hunk 1]
```diff
@@ -38,12 +38,11 @@ class Zmanim:  # pylint: disable=too-many-instance-attributes
     property.
     """
 
-    # pylint: disable=too-many-arguments, too-many-positional-arguments
```
Comment (tsvi) [inline hdate/zmanim.py:41]: I think you accidentally deleted this line

File: tox.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [tox]
-env_list = py{39,310,311,312,313}, lint
```
Comment (tsvi) [inline tox.ini:2]: Please don't change tox.ini

File: README.rst
Comment (tsvi) [inline README.rst:1]: Don't forget to fix the README.rst

**Review by tsvi — CHANGES_REQUESTED** [review]
Some minor comments. When everything passes, click the ready for review and I'll take a look at it again.

**Comment by DanBendavid** [comment]
Ouch ! I am not a "Real" developer. Thanks for your patience with me.
I need to add few tests for french.

**Comment by DanBendavid** [comment]
Look good . Now. 
Where can we have a open discussion for HA ?

**Comment by tsvi** [comment]
I'll need some more time to review this. Please ping me if I haven't replied to you by next week.

**Comment by DanBendavid** [comment]
I have finally modified the Holiday structure to split Language from definition entries.

**Comment by tsvi** [comment]
If you're already doing that, I was thinking of changing the multi language support using JSON data files.
Each file has as a name the language they support and then they use keys to get the description e.g.

`english.yaml`
```json
{
"holidays": { "rosh_hashana": "Rosh Hashanah", ...}
"months": { ... }
}
```

This allows to separate the logic from the data and is a good programming paradigm. If it's too complicated, we can leave it for a later PR

**Comment by DanBendavid** [comment]
> If you're already doing that, I was thinking of changing the multi language support using JSON data files. Each file has as a name the language they support and then they use keys to get the description e.g.
> 
> `english.yaml`
> 
> ```json
> {
> "holidays": { "rosh_hashana": "Rosh Hashanah", ...}
> "months": { ... }
> }
> ```
> 
> This allows to separate the logic from the data and is a good programming paradigm. If it's too complicated, we can leave it for a later PR

I understand the logic if we have to develop many languages. However i see 2  arguments for not doing 1 file per language :
 - i) very few more languages are required (spanish, german, russian, ???)
 - ii) keeping hebrew next to french or english is good for debugging

**Comment by tsvi** [comment]
a. Regarding the number of languages: I don't know. But 6 languages is reason enough to make it generic.
b. Since at the end we'll have a key for each name I don't see why it would be more complicated for debugging. In each file we have key value pairs. The key is a transliterated English name.

**Comment by DanBendavid** [comment]
Ok let’s start the work. But fist accept the PR. I will follow your guidance . You are the owner and the expert. Sent from my iPhone / DanOn 29 Nov 2024, at 12:58 AM, Tsvi Mostovicz ***@***.***> wrote:﻿
a. Regarding the number of languages: I don't know. But 6 languages is reason enough to make it generic.
b. Since at the end we'll have a key for each name I don't see why it would be more complicated for debugging. In each file we have key value pairs. The key is a transliterated English name.

File: hdate/date.py [hunk 3]
```diff
@@ -526,8 +589,24 @@ def hebrew_number(num: int, hebrew: bool = True, short: bool = False) -> str:
     return hstring
 
 
-def get_omer_string(omer: int) -> str:  # pylint: disable=too-many-branches
-    """Return a string representing the count of the Omer."""
+def get_omer_string(omer: int, language: str = "hebrew") -> str:
+    """Retourne une chaîne représentant le compte de l'Omer."""
```
Comment (tsvi) [inline hdate/date.py:593]: Please keep all the docstring in English.

File: hdate/date.py [hunk 4]
```diff
@@ -541,45 +620,202 @@ def get_omer_string(omer: int) -> str:  # pylint: disable=too-many-branches
...
-        if one:
-            omer_string += " ו"
+        omer_string += f"{ones[omer % 10]} עשר"
+    elif omer >= 10:
+        unit_part = ones[omer % 10]
+        if omer % 10:
+            unit_part += " ו"
+        omer_string += unit_part + tens[omer // 10]
+    else:
+        # Ne pas ajouter 'אחד' ou 'שנים' ici pour éviter la duplication
```
Comment (tsvi) [inline hdate/date.py:635]: Same for comments

File: hdate/date.py [hunk 5]
```diff
@@ -541,45 +620,202 @@ def get_omer_string(omer: int) -> str:  # pylint: disable=too-many-branches
...
+        15: "quinzième",
+        16: "seizième",
+        20: "vingtième",
+        30: "trentième",
+        40: "quarantième",
+    }
+
+    # Vérification que le nombre est dans la plage valide
+    if omer < 1 or omer > 49:
+        return "Le nombre doit être entre 1 et 49."
```
Comment (tsvi) [inline hdate/date.py:778]: No need to double check, this was checked in the general function

File: hdate/date.py [hunk 6]
```diff
@@ -94,6 +126,31 @@ def __ge__(self, other: "HDate") -> bool:
         """Implement the greater than or equal operator."""
         return not self < other
 
+    def get_hebrew_number(self, number: int, short: bool = False) -> str:
+        """Get the number representation based on the current languageuage."""
+        hstring: str = hebrew_number(num=number, language=self.language, short=short)
+        return hstring
+
```
Comment (tsvi) [inline hdate/date.py:133]: I understand the reason for wanting this helper method, but HDate is already big as is. Please remove. You can use `hebrew_number` directly.

File: hdate/date.py [hunk 7]
```diff
@@ -139,31 +196,40 @@ def _jdn(self) -> int:
...
-            for entry in entries
-        )
+        descriptions = []
+        for entry in entries:
+            # Access the language-specific description
+            description_language = getattr(
+                HOLIDAY_DESCRIPTIONS.get(entry.name), self.language, None
+            )
+            if description_language:
+                # Check if it's a DESC namedtuple with 'long' and 'short' attributes
```
Comment (tsvi) [inline hdate/date.py:227]: Thanks for dealing with that. Although I do plan getting rid of short descriptions.

File: hdate/date.py [hunk 8]
```diff
@@ -47,25 +63,41 @@ def __init__(
...
+            f"{day_prefix}{day_name} {day_number} "
+            f"{in_prefix}{month_name} {year_number}"
         )
-
+        # Handle Omer day
         if 0 < self.omer_day < 50:
-            result = f"{result} {hebrew_number(self.omer_day, hebrew=self.hebrew)}"
-            result = f"{result} {'בעומר' if self.hebrew else ' in the Omer'}"
+            omer_day_number = self.get_hebrew_number(self.omer_day)
+            # omer_string_long = get_omer_string(self.omer_day, self.language)
```
Comment (tsvi) [inline hdate/date.py:96]: Please don't leave commented code.

File: hdate/htables.py [hunk 9]
```diff
@@ -782,117 +607,181 @@ class HolidayTypes(Enum):
...
+                    holidays_list.append(description_language.long)
+                else:
+                    holidays_list.append(description_language)
+            else:
+                # If the specified language is not available, default to English
+                holidays_list.append(description_entry.english)
+        else:
+            # If no description is found, use the holiday name
+            holidays_list.append(h.name)
+
```
Comment (tsvi) [inline hdate/htables.py:780]: I suggest using a short method to define what should be returned, and then use a list comprehension to create the list:

```python
def holiday_name(entry, language):
    entry_language = getattr(entry, language, None)
    if entry_language:
        return entry_language.long if isinstance(entry_language, DESC) else entry_language
    return entry.english

holidays = [holiday_name(h, language) for h in HOLIDAYS]
holidays += ["Chanukah, Rosh Chodesh"]  # Need to fix here also support for other languages
```

File: hdate/zmanim.py [hunk 10]
```diff
@@ -59,9 +59,9 @@ def __init__(
         calculations.
         """
         self.location = location
-        self.hebrew = hebrew
-        self.candle_lighting_offset = candle_lighting_offset
-        self.havdalah_offset = havdalah_offset
+        self.language: str = language
+        self.candle_lighting_offset: int = candle_lighting_offset
+        self.havdalah_offset: int = havdalah_offset
```
Comment (tsvi) [inline hdate/zmanim.py:64]: Not needed

**Review by tsvi — COMMENTED** [review]
Some minor fixes necessary, please use english in comments and docstrings.

File: hdate/date.py [hunk 11]
```diff
@@ -47,25 +63,40 @@ def __init__(
...
+        day_prefix = self.DAY_PREFIXES.get(self.language, "")
+        in_prefix = self.IN_PREFIXES.get(self.language, "")
+
+        # Get day name
+        day_index = self.dow - 1  # Assuming dow is 1-based (Sunday=1)
+        day_language = getattr(htables.DAYS[day_index], self.language)
+        day_name = day_language.long  # Use 'long' or 'short' as needed
+
+        # Get day number representation
+        day_number = hebrew_number(self.hdate.day)
```
Comment (tsvi) [inline hdate/date.py:81]: Don't forget adding the language

File: hdate/date.py [hunk 12]
```diff
@@ -94,6 +125,31 @@ def __ge__(self, other: "HDate") -> bool:
         """Implement the greater than or equal operator."""
         return not self < other
 
+    def get_hebrew_number(self, number: int, short: bool = False) -> str:
```
Comment (tsvi) [inline hdate/date.py:128]: No need for this method and mentioned

File: hdate/tekoufa.py
Comment (tsvi) [inline hdate/tekoufa.py:1]: Please remove this file from this PR.

**Review by tsvi — CHANGES_REQUESTED** [review]

File: hdate/date.py (hunk 4 above)
Comment (DanBendavid) [reply hdate/date.py]: ok

File: hdate/date.py (hunk 5 above)
Comment (DanBendavid) [reply hdate/date.py]: ok

File: hdate/date.py (hunk 8 above)
Comment (DanBendavid) [reply hdate/date.py]: ok

File: hdate/zmanim.py (hunk 10 above)
Comment (DanBendavid) [reply hdate/zmanim.py:64]: When removing : int I have many error "no-untyped-def"

File: hdate/date.py (hunk 12 above)
Comment (DanBendavid) [reply hdate/date.py]: Ok

**Review by tsvi — APPROVED** [review]
Thank you very much. Fixed some minor issues. Approved 👍🏻

**Comment by DanBendavid** [comment]
Thank you Tsvi for your patience with me. I have struggled a lot with Pylint !!! 

 

Now for HA, as I was telling you, my wish is to be able to combine both Hebrew with English or Hebrew with French on few Sensors ( Paracha, Holidays, Omer, Daf, Date etc….)

We have 2 options : 1) We duplicate few sensors 

or we add the second language as an attribute in sensors. 

 

Let me know you view.

Dan

  

 

De : Tsvi Mostovicz ***@***.***> 
Envoyé : dimanche 8 décembre 2024 20:14
À : py-libhdate/py-libhdate ***@***.***>
Cc : Dan BENDAVID ***@***.***>; Author ***@***.***>
Objet : Re: [py-libhdate/py-libhdate] Multilanguage (PR #144)

 

@tsvi approved this pull request.

Thank you very much. Fixed some minor issues. Approved 👍🏻

**Comment by tsvi** [comment]
I think the best way to approach this is using the following:

https://developers.home-assistant.io/docs/internationalization/core/#state-of-entities

What I would like is to have each sensor return the Hebrew value always as
an attribute. The sensor value will be the language chosen by the frontend
this way:
a. We're compatible with home assistant development guidelines
b. The user can always create a Hebrew template sensor from the attribute
if they wish to have that.

To do this we best move to use a JSON (I'll try to implement this
tomorrow). Which we then can pull into the home assistant project.

**Comment by DanBendavid** [comment]
Perfect for me. Let me know how i can help you. 

 

De : Tsvi Mostovicz ***@***.***> 
Envoyé : dimanche 8 décembre 2024 23:00
À : py-libhdate/py-libhdate ***@***.***>
Cc : Dan BENDAVID ***@***.***>; Author ***@***.***>
Objet : Re: [py-libhdate/py-libhdate] Multilanguage (PR #144)

 

I think the best way to approach this is using the following: 

https://developers.home-assistant.io/docs/internationalization/core/#state-of-entities 

What I would like is to have each sensor return the Hebrew value always as 
an attribute. The sensor value will be the language chosen by the frontend 
this way: 
a. We're compatible with home assistant development guidelines 
b. The user can always create a Hebrew template sensor from the attribute 
if they wish to have that. 

To do this we best move to use a JSON (I'll try to implement this 
tomorrow). Which we then can pull into the home assistant project.

**Pull request merged**